feat(datagrid): virtual scroll

valentin-mladenov opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: CDE-1749

## What is the new behavior?

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

dtsanevmw commented on projects/angular/src/data/datagrid/datagrid.ts:
```diff
@@ -274,7 +278,19 @@ export class ClrDatagrid<T = any> implements AfterContentInit, AfterViewInit, On
-          row ? this.detailService.open(row.item, row.detailButton.nativeElement) : this.detailService.close();
+          if (row) {
+            this.detailService.open(row.item, row.detailButton.nativeElement);
+          } else if (!this.hasVirtualScroller) {
```
We need to keep the old condition where row is not available.
```suggestion
          } else if (!this.hasVirtualScroller || !row) {
```

valentin-mladenov replied:
Corrected.

dtsanevmw commented on projects/angular/src/data/datagrid/datagrid.html:
```diff
@@ -22,7 +22,7 @@
                     (keydown.space)="allSelected = !allSelected; $event.preventDefault()"
                   >
-                    <div class="clr-checkbox-wrapper">
+                    <div *ngIf="!hasVirtualScroller" class="clr-checkbox-wrapper">
```
Clicking the header throws error from 'datagrid-selection-cell.directive.ts' because we are still tracking click events by the .div abbove

valentin-mladenov replied:
Corrected.

mivaylo commented on projects/angular/src/data/datagrid/datagrid-virtual-scroll.directive.spec.ts:
```diff
@@ -0,0 +1,279 @@
+export interface Row {
+  index: number;
+  expanded: boolean;
+  cells: any[];
```
As discussed here could be used some typization instead of any

I think we discussed that example:

`interface cells {
   [key: string]: string;
}`

valentin-mladenov replied:
Corrected.

dtsanevmw commented on projects/angular/src/data/datagrid/datagrid-virtual-scroll.directive.spec.ts:
```diff
@@ -0,0 +1,229 @@
+        await fixture.whenStable();
+        await sleep();
+        // active checkbox input with ID clr-dg-row-cb383
+        console.log(3, document.activeElement);
+        console.log(3, grid.querySelectorAll('[type=checkbox]')[0]);
```
Left over code.

valentin-mladenov replied:
removed .

mivaylo commented on .storybook/helpers/elements.data.ts:
```diff
@@ -5,130 +5,135 @@
-  { name: 'Yttrium', symbol: 'Y', number: 39, electronegativity: 1.22 },
-  { name: 'Zinc', symbol: 'Zn', number: 30, electronegativity: 1.65 },
-  { name: 'Zirconium', symbol: 'Zr', number: 40, electronegativity: 1.33 },
+  { expanded: false, name: 'Actinium', symbol: 'Ac', number: 89, electronegativity: 1.1 },
```
As discussed this new optional field could omitted here, by just using "!!element?.expanded" when the new field is used. But it's just personal opinion - this means we are setting the default value to false (which maybe does not correspond to our need? - I am not sure about that)

valentin-mladenov replied:
Corrected.

dtsanevmw commented on projects/angular/src/data/datagrid/datagrid-virtual-scroll.directive.spec.ts:
```diff
@@ -0,0 +1,229 @@
+        await sleep();
+        await fixture.whenStable();
+        await sleep();
+        console.log(4, document.activeElement);
+        console.log(4, grid.querySelectorAll('[type=checkbox]')[0]);
```
Left over code.

valentin-mladenov replied:
Removed in latest commits.

dtsanevmw commented on projects/angular/src/data/datagrid/datagrid-virtual-scroll.directive.spec.ts:
```diff
@@ -0,0 +1,229 @@
+        expect(document.activeElement).toEqual(grid.querySelectorAll('[type=checkbox]')[22]);
+
+        grid.dispatchEvent(new KeyboardEvent('keydown', { code: 'PageDown' }));
+        await sleep();
+        await fixture.whenStable();
+        await sleep();
```
One sleep with additional time is enough.
```suggestion
        await sleep(50);
        await fixture.whenStable();
```

valentin-mladenov replied:
It's different in latest commits.

dtsanevmw commented on projects/angular/src/data/datagrid/utils/key-navigation-grid.controller.ts:
```diff
@@ -88,6 +89,12 @@ export class KeyNavigationGridController implements OnDestroy {
           }
         });
 
+      fromEvent(this.grid, 'wheel')
+        .pipe(takeUntil(this.destroy$))
+        .subscribe(() => {
+          this.removeActiveCell();
+        });
+
```
Why do we need that? It's removing the focus if you scroll with the mouse couple of times and the row is still in view.

valentin-mladenov replied:
It will be triggered only after data update. If the data is not changed the focus will remain.

dtsanevmw replied:
It's tracking 'wheel' events so it's being removed if you focus some cell and start scrolling -- check video
https://github.com/vmware-clarity/ng-clarity/assets/127101685/6613cb6a-734a-4580-ac66-a9fd726bdb65

valentin-mladenov replied:
This issue will be addressed in [CDE-2043](https://jira.eng.vmware.com/browse/CDE-2043)

mivaylo commented on projects/angular/src/data/datagrid/datagrid.spec.ts:
```diff
@@ -874,12 +874,18 @@ export default function (): void {
+        // focus at top datagrid radio input
         grid.dispatchEvent(new KeyboardEvent('keydown', { code: 'PageUp' }));
-        expect(document.activeElement).toEqual(cells[0]);
+        expect(document.activeElement).toEqual(cells[3].querySelector('[type=radio]'));
```
As discsussed some comments for why we use this number 3 here could be useful

valentin-mladenov replied:
Comments added.

mivaylo reviewed (changes requested):
Left a few comments/ideas that could be nice to have IMHO as discussed

valentin-mladenov commented on projects/angular/src/data/datagrid/datagrid-virtual-scroll.directive.spec.ts:
```diff
@@ -0,0 +1,229 @@
+        grid.dispatchEvent(new KeyboardEvent('keydown', { code: 'PageUp' }));
+        await sleep();
+        await fixture.whenStable();
+        await sleep();
```
It's different in latest commits.

dtsanevmw commented on projects/angular/src/data/datagrid/datagrid-virtual-scroll.direcive.ts:
```diff
@@ -0,0 +1,332 @@
+  private mutationChanges: MutationObserver = new MutationObserver((mutations: MutationRecord[]) => {
+    mutations.forEach((mutation: MutationRecord) => {
+      // it is possible this to be called twice because the old class is removed and the new added
+      if ((mutation.target as HTMLElement).classList.contains('datagrid-compact')) {
```
Add condition to check if the value is already the same to skip the itemSize setter and updating the virtual scroll values constantly.
```suggestion
      if ((mutation.target as HTMLElement).classList.contains('datagrid-compact')  && this.itemSize > 24) {
```

valentin-mladenov replied:
Corrected in commit 8f155e7.

dtsanevmw commented on projects/angular/src/data/datagrid/datagrid.ts:
```diff
@@ -274,7 +278,19 @@ export class ClrDatagrid<T = any> implements AfterContentInit, AfterViewInit, On
+        }
+
+        // retain active cell
+        const active = this.keyNavigation.getActiveCell();
+        if (active) {
+          this.zone.runOutsideAngular(() => {
+            setTimeout(() => this.keyNavigation.setActiveCell(active));
+          });
```
When this case would happen? Because we are removing the activeCell on mouse wheel and I see it's null most of the time here.

valentin-mladenov replied:
This is to keep the focus when navigating with `PageUp` and `PageDown` buttons. 

Comment added in commit 8f155e7.

dtsanevmw reviewed (commented):
- I see in storybook we are using a small list of items but wouldn't it be better to use larger amount of items there?
- Can we add a story with detail because I see that there are some overflow changes there and can be tested?
- Explore different name for the directive - 'ClrVirtualScroll'
- Directive file name has a typo '.direCIve' needs to be '.directive.'

valentin-mladenov commented:
@dtsanevmw responding to your list of requests

- I see in storybook we are using a small list of items but wouldn't it be better to use larger amount of items there like in the demo app?
The list used have enough items to represent a functioning virtual scroller, if needed wider list of items can be added at later state with different PR.
- Can we add a story with detail because I see that there are some overflow changes there and can be tested?
Added in commit: 1d38308.
- Explore different name for the directive - 'ClrVirtualScroll'
New naming convention added in commits: dbc7fd0 and 9c522a0.
- Directive file name has a typo '.direCIve' needs to be '.directive.'
Corrected in commit: dbc7fd0.

valentin-mladenov commented on .storybook/stories/datagrid/datagrid-virtualscroll.stories.ts:
```diff
@@ -0,0 +1,239 @@
+    // inputs
+    clrDgSelected: { control: { disable: true } },
+    clrDgSingleSelected: { control: { disable: true } },
+    // outputs
```
One of the `// output` comments is remove in commit 1d38308
